data/abi/abi_encode.go

```diff
@@ -268,203 +303,168 @@ func tupleEncoding(v Value) ([]byte, error) {
 	return encoded, nil
 }
 
-// Decode takes an ABI encoded byte string and a target ABI type,
-// and decodes the bytes into an ABI Value.
-func Decode(valueByte []byte, valueType Type) (Value, error) {
-	switch valueType.abiTypeID {
-	case Uint:
-		if len(valueByte) != int(valueType.bitSize)/8 {
-			return Value{},
-				fmt.Errorf("uint%d decode: expected byte length %d, but got byte length %d",
-					valueType.bitSize, valueType.bitSize/8, len(valueByte))
-		}
-		uintValue := new(big.Int).SetBytes(valueByte)
-		return MakeUint(uintValue, valueType.bitSize)
-	case Ufixed:
-		if len(valueByte) != int(valueType.bitSize)/8 {
-			return Value{},
-				fmt.Errorf("ufixed%dx%d decode: expected length %d, got byte length %d",
-					valueType.bitSize, valueType.precision, valueType.bitSize/8, len(valueByte))
-		}
-		ufixedNumerator := new(big.Int).SetBytes(valueByte)
-		return MakeUfixed(ufixedNumerator, valueType.bitSize, valueType.precision)
+// compressBools takes a slice of interface{} (which can be casted to bools) length <= 8
+// and compress the bool values into a uint8 integer
+func compressBools(boolSlice []interface{}) (uint8, error) {
+	var res uint8 = 0
+	if len(boolSlice) > 8 {
+		return 0, fmt.Errorf("compressBools: cannot have slice length > 8")
+	}
+	for i := 0; i < len(boolSlice); i++ {
+		temp, ok := boolSlice[i].(bool)
+		if !ok {
+			return 0, fmt.Errorf("compressBools: cannot cast slice element to bool")
+		}
+		if temp {
+			res |= 1 << uint(7-i)
+		}
+	}
+	return res, nil
+}
+
+// Decode is an ABI type method to decode bytes to go values from ABI encoding rules
+func (t Type) Decode(encoded []byte) (interface{}, error) {
+	switch t.abiTypeID {
+	case Uint, Ufixed:
+		if len(encoded) != int(t.bitSize)/8 {
+			return nil,
+				fmt.Errorf("uint/ufixed decode: expected byte length %d, but got byte length %d", t.bitSize/8, len(encoded))
+		}
+		return new(big.Int).SetBytes(encoded), nil
```

Comment:
I wonder if we should return the types `uint8`, `uint16`, `uint32`, `uint64` if the ABI type precision can fit into them. Something like:

| ABI precision  | Go type |
| ------------- | ------------- |
| 8  | `uint8` |
| 16  | `uint16` |
| 24 - 32 | `uint32` |
| 40 - 64 | `uint64` |
| > 64 | `*big.Int` |

Reply:
It is doable, and I think this is a very nice improvement, but the testing will have more scripts to support different types. Can't do it today, leave it tomorrow...